### DISCUSSION TITLE
Preventing source plug-ins from added as depencies in the MANIFEST.MF

### DISCUSSION THREAD

**Discussion by vogella** [eclipse-pde/eclipse.pde#51] [Ideas]
In my plugin and RCP training sessions participants frequently add "sources" as dependencies in the MANIFEST and therefore their products do not start anymore as they are not part of their product / features. 

Can we avoid this? Maybe we can just filter sources from the list of plugins which can be added to the MANIFEST.MF? cc @HannesWell who recently update the source generation in multiple places.

**Comment by HannesWell** [comment]
At least Source-Bundles build by Tycho have a `Eclipse-SourceBundle` header in their MANIFEST and IIRC PDE requires that in order to be able to locate sources of bundles. This would be a reliable test that does not rely on naming conventions.
But I'm not sure if this also applies to the sources of all Maven-artifacts (it would be great, but I don't think this is the case for all that have a OSGi-compliant MANIFEST). However @laeubi recently created a Tycho-Plugin to simplify that: https://github.com/eclipse/tycho/blob/master/RELEASE_NOTES.md#automatic-generation-of-pde-source-bundles-for-pom-first-bundles.

And once we have a reliable test if a bundle is a source-bundle it should be fairly easy to filter out source bundles in the `Add Required Plug-ins` dialog. Additionally PDE could issue an error with explanatory message (whose severity can be configured) if it detects a required source-bundle in existing manifests.

Nevertheless in general it is recommended to use Import-Package instead of Require-Bundle (which is deprecated according to the OSGi-spec). For example the transition to OSGi-Bundles from Maven-Central would be simpler if Eclipse would use only Import-Package. The Plug-ins embedding those source could then be removed immediately. But before we evangelize this more PDE should improve its support for that (but that's another discussion).

**Reply by laeubi** [reply]
> Additionally PDE could issue an error with explanatory message (whose severity can be configured) if it detects a required source-bundle in existing manifests.

Source bundles are regular bundles so it is not an error.

> And once we have a reliable test if a bundle is a source-bundle it should be fairly easy to filter out source bundles in the `Add Required Plug-ins` dialog

We better error the on using require-bundle at all ;-)

**Reply by HannesWell** [reply]
> > Additionally PDE could issue an error with explanatory message (whose severity can be configured) if it detects a required source-bundle in existing manifests.
> 
> Source bundles are regular bundles so it is not an error.

That's right. Then it should be a warning, although people are too used to warnings and therefore often oversee them respectively have too many in the workspace that new ones are not recognized. But making them used to errors is probably also not a good idea.  

> > And once we have a reliable test if a bundle is a source-bundle it should be fairly easy to filter out source bundles in the `Add Required Plug-ins` dialog
> 
> We better error the on using require-bundle at all ;-)

I do not disagree but its another discussion. :)

**Reply by laeubi** [reply]
@HannesWell just for completeness PDE requires the `Eclipse-SourceBundle` header and the maven target integration adds missing headers if required.
If you include maven dependencies directly you can't use them as source bundles (unless they already have headers), but if you **produce** maven-first bundles you can generate a source bundle that contains the headers. 

Beside that, PDE won't strictly require the header and actually the target-platfrom has a method to test if an *arbitrary* bundle is a source bundle but it seems not be implemented down further in PDE itself to use that method.

**Comment by vogella** [comment]
> But I'm not sure if this also applies to the sources of all Maven-artifact

Some support is better than no support, so I will create a issue and plan to look into it. Thanks Hannes.

**Comment by laeubi** [comment]
> Can we avoid this? Maybe we can just filter sources from the list of plugins which can be added to the MANIFEST.MF?

The question is: Why they do that? This is not an error and maybe even desired, who knows?

**Reply by vogella** [reply]
> > Can we avoid this? Maybe we can just filter sources from the list of plugins which can be added to the MANIFEST.MF?
> 
> The question is: Why they do that? This is not an error and maybe even desired, who knows?

Not desired, they just copy the plug-in into the dialog, which filters already existing entries. Therefore the only entry is the source entry which they pick therefore.

**Reply by HannesWell** [reply]
I would say by accident or because they don't know better or think it could be necessary to make them available (users do strange things if they can and do not fully understand the problem, I know this from myself).

Is there any scenario where it is useful to require a source bundle?

**Reply by laeubi** [reply]
But here, again, why? It does not make sens to "copy something" and inset arbitrary things to the manifest if it is not required... I really don't see to put effort here for people that are to lazy and blindlessly adding things...

**Reply by vogella** [reply]
> I really don't see to put effort here for people that are to lazy and blindlessly adding things

I see value in helping new users of the platform to avoid starting errors.

**Reply by laeubi** [reply]
If we want to help users, then we should not recommend them Require-Bundle regardless if SWT might make it "inconvenient" ... that are not "starting error" that's "try-and-error" to just trying to add arbitrary things, e.g. just in case the dialog is empty won't such a user blame this to a PDE error?

A much better approach would be to just show even bundles that are already added e.g. in a gray/disabled out manner and maybe appending "already imported" or something alike.

**Reply by vogella** [reply]
@laeubi feel free to open new issues for non related items.

Again:

> We better error the on using require-bundle at all ;-)

I also  do not disagree but its another discussion. :)

**Reply by laeubi** [reply]
This is related, I don't think we should filter things out to help people that don't know the have imported and accidentally choose something, we should choose a solution that helps the user making correct choices.

So the problem is not that source bundles are shown, but that users try to add things and if they don't see it choose something that sounds similar and we should then solve this and not decide that showing something confuses the users (its the thing they don'T see actually confusing them...)

**Reply by vogella** [reply]
This issue is just about source bundles allowed be added a dependencies, for which we do not have a valid use case. If we have no use case for that, we should not allow it.

**Reply by laeubi** [reply]
Why is it not a valid use case? You can require any bundle and a source-bundle is a bundle...

**Reply by vogella** [reply]
> Why is it not a valid use case? You can require any bundle and a source-bundle is a bundle...

Let me know if you have a valid use case other than, "hey it is currently possible". The purpose to add a requirement it to use it in your code or to require it at runtime.

**Reply by vogella** [reply]
Even if the wizard would not allow to add source bundles, the user still could add them manually in the MANIFEST.MF if he likes.

**Reply by vogella** [reply]
Also, in case you find a use case for require soure plugins, this would be mean we should (joke is coming, don't take it serious), deprecate import package dependencies, as itdoes not support source package dependencies.

**Reply by laeubi** [reply]
If you want enforce source is delivered with your bundle (e.g. legal reasons) one might choose that aproach, or you might choose to access resources from the source bundle, neither is what I regular use...

But still this does not solve anything I think if we assume that the user is *confused* because he would equally confused by having an empty dialog or having e.g. other bundles just wit the same prefix.

**Reply by vogella** [reply]
> neither is what I regular use...

Never seen this usage and we should guide people not to rely on it otherwise we can never move them to use import packages. :-)

**Reply by HannesWell** [reply]
> Even if the wizard would not allow to add source bundles, the user still could add them manually

The wizard is not forbidding to add source bundles, it just not proposing to add source bundles then.
From my understanding when the wizard shows a proposal, it means PDE thinks it is a good idea to do something, so by enabling a certain proposal we encourage users to do what the proposal is proposing. So by listing the source bundles we effectively encourage people to add them and I don't think that is what we want (in 99%).
It is similar to the "Compute" Dependencies/ "Recompute when feature plugins change" Buttons in the Feature-Editor. When there is something proposed users will start using it. From my experience, if in doubt users tend to do/include more than necessary. It often requires deep knowledge to keep the configuration minimal. But usually it is better to be minimal.

And adding a warning to existing source-bundle requirements could help the user to re-consider its previous decision. If this is what a user really wants the warning could be downgraded to info/ignore. But even then we have already gained much because a user then knows (more) about the purpose/nature of source bundles and can make an informed decision (which is IMHO very important).

**Reply by laeubi** [reply]
> The wizard is not forbidding to add source bundles, it just not proposing to add source bundles then.

If you don't offer them, how do you add them (when not editing the manifest as text)?

> From my understanding when the wizard shows a proposal, it means PDE thinks it is a good idea to do something,

I don't think so, it do simply shows all bundles and you can filter by id, there is no "X is better than Y" ...

> So by listing the source bundles we effectively encourage people to add them

I don't think so that would mean we encourage people to add all bundles of a workspace by default as this is what you get when opening the dialog.

> if in doubt users tend to do/include more than necessary.

If in doubt he better use automatically managed dependencies tab then...

> But even then we have already gained much because a user then knows (more) about the purpose/nature of source bundles and can make an informed decision

Honestly I won't invest any time in adding/supporting such features, this is completely just to "help" people the wrong way as I noted above, the problem is not that source bundles are shown (we might simply add a checkbox [ ] hide source bundles) the problem is that bundles are not shown that are already added to the manifest and that's hard to understand (e.g. also bundles are hidden where there is alredy an import package) because you often think if you have typed something wrong. So we better do not hide but mark items that are already added... even if I select the bundle, PDE could simply do nothing if it is already there, who cares?

But assuming that it is **common** that users add source bundles by accident and we must do warn them is completely strange ... this would more indicate that users have absolutely no clue and adding **anything** they got offered for now good... PDE is for developers and we can at laest assume a users has a basic understanding what he/she wants to archive.

If one really likes to improve things for developers there are more valuable things like:

- Finally get rid of the the Manifest editor assuming that each javax. package is auto imported
- Show if packages/bundles are imported but never used
- Support adding proper version ranges
- Supporting split-packages as eclipse itself is the most common source for them
- ....

**Reply by vogella** [reply]
@laeubi again you trying to discuss other things here. If you do not plan to invest time in this, that is fine, I did not ask you to work on this. Maybe in this case also stop spending time in arguing why non related fuctionality is not good. This "I use an issue to discuss things which are important to me" is not adding value (for me) and I suggested already twice that you stop doing that.

**Reply by laeubi** [reply]
@vogella this is an open discussion so I think you have to scope with that. And thats what Discussions are for, looking at different angels to the same problem and maybe suggesting an alternative approach that solves the **use case** or **user story**.
If you like to discuss about a concrete **solution**  to a problem its generally better to open a PR with actual code changes ...

**Reply by vogella** [reply]
> @vogella this is an open discussion so I think you have to scope with that

:-) you are funny.

**Reply by laeubi** [reply]
OSS **is** fun isn't it :-) If I'm doing what other telling me it is most often payed work and not OSS and not always fun ...

**Reply by vogella** [reply]
> OSS **is** fun isn't it :-) If I'm doing what other telling me it is most often payed work and not OSS and not always fun ...

Indeed. I'm also in the luxary position of getting almost never paid for OS work,  just image how stressful it would be to work on OS and get money from it. :-)

**Reply by laeubi** [reply]
> Indeed. I'm also in the luxary position of getting almost never paid for OS work

You probably could ask your customer of that training sessions if the want to pay you to implement (including tests, documentation PRs _discussions_) this feature... or if that increases the learning curve not adding source bundles at the first place :-D

**Reply by vogella** [reply]
Every *new* group of people makes this error. So if I implement this it is mainly for me, saving the time to explain new plugin developers that adding source bundles as dependencies is wrong. It is one of these things you (hopefully) do only wrong once.

**Comment by vogella** [comment]
> Nevertheless in general it is recommended to use Import-Package instead of Require-Bundle

IMHO this depends. If a plug-in requires tons of packages,  e.g. like SWT it is so much easier to use plug-in dependencies compared to import packages as you would have to add a lot of package to use SWT which makes it time consuming and hard to read.
Also we have no alternative SWT implementations floating around using import package for SWT is not adding value.

**Reply by laeubi** [reply]
You don't need to add *all* packages anyways and PDE suggest you to import them as you have missing imports.

**Comment by vogella** [comment]
But lets not start a import package vs. required bundles here. This is about source bundles. :-)

**Comment by vogella** [comment]
https://github.com/eclipse-pde/eclipse.pde/issues/52
